discovery_engine_core/web-api2/src/personalization/routes.rs

```diff
@@ -40,33 +58,203 @@ pub(super) fn configure_service(config: &mut ServiceConfig) {
     config.service(scope);
 }
 
-//FIXME use actual body
-#[derive(Deserialize, Debug)]
-struct UpdateInteractions {}
+/// Represents user interaction request body.
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct UpdateInteractions {
+    pub(crate) documents: Vec<UserInteractionData>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct UserInteractionData {
+    #[allow(dead_code)]
+    #[serde(rename = "id")]
+    pub(crate) document_id: DocumentId,
+    #[allow(dead_code)]
+    #[serde(rename = "type")]
+    pub(crate) interaction_type: UserInteractionType,
+}
 
 async fn update_interactions(
-    config: Data<Config>,
+    state: Data<AppState>,
     user_id: Path<UserId>,
-    interactions: Json<UpdateInteractions>,
+    Json(interactions): Json<UpdateInteractions>,
 ) -> Result<impl Responder, Error> {
-    dbg!((config, user_id, interactions));
-    if true {
-        Err(Unimplemented {
-            functionality: "/users/{user_id}/interactions",
-        })?;
+    state.db.user_seen(&user_id).await?;
+
+    let ids = interactions
+        .documents
+        .iter()
+        .map(|document| &document.document_id)
+        .collect_vec();
+    let documents = state.elastic.get_documents_by_ids(&ids).await?;
+    let embeddings = documents
+        .into_iter()
+        .map(|document| (document.id, document.embedding))
+        .collect::<HashMap<_, _>>();
+
+    for document in interactions.documents {
+        match document.interaction_type {
+            UserInteractionType::Positive => {
+                //TODO for some reason this was returning a BAD_REQUEST error????
+                state
+                    .db
+                    .update_positive_cois(&document.document_id, &user_id, |positive_cois| {
+                        state.coi.log_positive_user_reaction(
+                            positive_cois,
+                            &embeddings[&document.document_id],
+                        )
+                    })
+                    .await?;
+            }
+        }
+    }
+
+    Ok(HttpResponse::NoContent())
+}
+
+/// Represents personalized documents query params.
+#[derive(Debug, Clone, Deserialize)]
+struct PersonalizedDocumentsQuery {
+    count: Option<usize>,
+}
+
+impl PersonalizedDocumentsQuery {
+    fn document_count(&self, config: &PersonalizationConfig) -> usize {
+        self.count
+            .map(|count| count.clamp(0, config.max_number_documents))
+            .unwrap_or(config.default_number_documents)
```

Comment:
:grey_exclamation: Clients and server can get out of sync, so it's better to clamp the count instead of erroring or else there is a risk of a client stopping working if we ever decrease `max_number_documents`.

Also made that a configuration option instead of hard-coding it to 100.